Feat: Add Cookbook for Content-Based Recommendations with Gemini & Qdrant

andycandy opened the pull request:
**Status: Work in Progress / Seeking Feedback**

This PR presents the initial implementation of the Gemini+Qdrant recommendation cookbook outlined in #690. While the core functionality is present, I am actively seeking feedback and collaborating with reviewers to refine the notebook's structure, clarity, and explanations before final merging. Please feel free to add comments and suggestions!

## Description
This PR introduces a new cookbook example demonstrating how to build a scalable, content-based recommendation system using Google's Gemini API for semantic embeddings and Qdrant as an efficient vector database.

## Motivation
As outlined in #690 , existing examples often use toy datasets or focus on narrow use cases. This cookbook aims to provide a more practical, near-production-level demonstration showing how Large Language Model (LLM) embeddings, specifically from Gemini, can power recommendations across potentially diverse media types without relying on user interaction history (addressing cold-start scenarios).

## Solution:
A new Jupyter notebook, [Movie_Recommendation](./examples/qdrant/Movie_Recommendation.ipynb), has been added. This notebook guides the user through the following steps:

1.  **Setup:** Installs necessary libraries (`google-genai`, `qdrant-client`, `pandas`), configures the Gemini API client, and initializes the Qdrant client.
2.  **Data Loading & Preparation:**
    *   Loads a sample of the TMDB Movies Dataset. 
    **Note:** The notebook uses a *sample* (e.g., 5000 items) for demonstration purposes due to the scale of the full dataset and API costs associated with embedding millions of items. Clear comments explain this and how to adapt for the full dataset.
    *   Selects relevant features (`title`, `overview`, `genres`, `keywords`, `tagline`, `release_date`).
    *   Performs data cleaning: handles missing titles, fills missing optional text fields, and extracts the release year.
    *   Constructs a combined `text_for_embedding` string for each movie, leveraging available metadata.
3.  **Embedding & Indexing with Qdrant:**
    *   Defines a function (`get_embeddings_batch`) to efficiently generate embeddings for batches of text using the Gemini `embedding-001` model, including retry logic.
    *   Creates a Qdrant collection with appropriate vector parameters (size 768, Cosine distance).
    *   Implements a **batched upsert process**:
        *   Iterates through the data sample in batches.
        *   Generates embeddings for each batch via the Gemini API.
        *   Creates Qdrant `PointStruct` objects containing the vector, a unique ID (`movie_id`), and a structured `payload` (movie metadata like title, genre, year, overview).
        *   Upserts these points into the Qdrant collection in batches for efficiency.
    *   Verifies the number of points indexed in Qdrant.
4.  **Querying & Recommendation:**
    *   Defines a `recommend_movies` function that:
        *   Takes a natural language query (e.g., movie title, theme description).
        *   Generates an embedding for the query using Gemini (`task_type="RETRIEVAL_QUERY"`).
        *   Performs a similarity search against the indexed movie vectors in Qdrant using `client.search`.
        *   Returns the top K most similar movies, including their metadata (payload) and similarity scores.
    *   Includes example queries to demonstrate the recommendation functionality.




## Disclaimer

This PR uses the [TMDb Movie Dataset](https://www.kaggle.com/datasets/tmdb/tmdb-movie-metadata) for non-commercial purposes only, as per its licensing terms.

- The dataset is licensed under **CC BY-NC 4.0**, which restricts usage to non-commercial projects.  
- Proper attribution has been provided to TMDb as required by the license.

By submitting this PR, I confirm that:
1. This dataset is used solely for educational and demonstration purposes.
2. Any further use of this dataset must comply with its licensing terms, and users are responsible for ensuring compliance.

andycandy commented:
@markmcd Here is my first rough draft, could you go through it once

Giom-V commented on examples/qdrant/Movie_Recommendation.ipynb:
```diff
@@ -0,0 +1,919 @@
+{
```
<p>More explanation as to why.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Giom-V replied:
<p>I think you need to reintroduce Qdrant and explain in details what you are doing (creating a client). Also a dumb question everybody will wonder: where is it stored?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Giom-V replied:
<p>Explain why you need to create a payload</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/#comment-2126127016'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Giom-V replied:
<p>Since this is quite a long cell, explain both in text and with comments in the code what is happening.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Giom-V commented:
@andycandy Overall I like the example, and I think it would be a great addition to the cookbook. I just think it needs way more explanations (but that was a first draft so that was expected).
Don't also forget to create a readme in the folder, update the one in examples/, and maybe update the "what's next" section of the embedding notebooks (and maybe other related examples) to mention this one.

andycandy commented:
Thanks! I've completed all the required updates: added the README in the qdrant/ folder, updated the main examples/ README, and included this notebook in the "What's next" and related examples sections of the relevant notebooks, n also expanded the explanations throughout the notebook for clarity.

Giom-V commented on examples/qdrant/Movie_Recommendation.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<p>Can you add links to the source?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Giom-V replied:
<p>I think you should explain why you want t do that (the dataset is huge, you don't need it all, so you want to get a feel of what's in there to select only what you are interested in)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Giom-V replied:
<pre class="ql-syntax" spellcheck="false">Line #40.    print(f"Batch embedding function 'get_embeddings_batch' defined using model: {MODEL_FOR_EMBEDDING}")
</pre><p>I don't think this line helps in any way. I think you should instead show an example of what an embedding is.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Giom-V commented on examples/qdrant/Movie_Recommendation.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<p>Explain a bit more why in your case you only want those columns. (I'm surprised you're saying you're removing the movie title.).</p><p>EDIT: actually I think the main issue is that your sentence is hard to understand. Are you keeping movie title, overview and genres or not? And why?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

andycandy replied:
<p>Yeah, I didn't frame it quite well. I referred to rows as entries. I am dropping rows that are missing titles or lack both overview and genre because such entries can not contribute to any value in our dataset.</p>

Giom-V commented on examples/qdrant/Movie_Recommendation.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<p>I'd say "To improve efficiency during development and keep this notebook easy to run" or something like that. Also instead of saying "it's common to do X" it would be better to say "you will want to be able to iterate quickly and keep the resource consumption to a minimum, so instead of using the full database..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

andycandy replied:
<p>Got it!</p>

Giom-V commented on examples/qdrant/Movie_Recommendation.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<p>It's better to write "the <code>create_payload</code> function prepares..." that "the function below...". It think it needs to be improved all over the notebook. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

andycandy replied:
<p>Yeah, that makes sense. I'll change this.</p>

Giom-V commented on examples/qdrant/Movie_Recommendation.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #24.            search_result = client.search(
</pre><p> </p><p>For clarity since there are both a qdrant and a gemini clients, I think this one should be renamed <code>qdrant_client</code>. If you want to call one only <code>client</code> it needs to be the genai one as in the rest of the cookbook. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google-gemini/cookbook/pull/700/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

andycandy replied:
<p>Understood!</p>

andycandy commented on examples/qdrant/Movie_Recommendation.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<p>Yeah, that's much better.</p>

andycandy commented on examples/qdrant/Movie_Recommendation.ipynb:
```diff
@@ -0,0 +1,1135 @@
+{
```
<p>Yeah, that's true.</p>
